### PR TITLE
[move-lang] allow _x to bind but do not warn as unused local

### PR DESCRIPTION
**Pull request by mengxu-fb**
## Motivation

`_` patterns are used to not introduce a binding, e.g. `let (x, _, z) = (0, 1, 2);`
- In Rust, if you do `_y` instead of `_`, e.g. `let (x, _y, z) = (0, 1, 2);` This introduces a binding `_y = 1` but it will not complain if you don't use it
- In Move currently `_y` is just sugar for `_` which isn't really helpful

The fix:
- make `_y` not remap to `_`, this will automatically fix being able to use it
- add an exclusion for underscore prefixed locals in the unused local warning

## Test Plan

`cargo xtest`

### CODE REVIEW

**Comment by mengxu-fb** [comment]
Updated test, `almost_invalid_local_name.move` is still valid. The invalid test is moved to `invalid_local_name.move`.

File: language/move-lang/src/cfgir/liveness/mod.rs [hunk 1]
```diff
@@ -294,12 +294,14 @@ mod last_usage {
                     match display_var(v.value()) {
                         DisplayVar::Tmp => (),
                         DisplayVar::Orig(v_str) => {
-                            let msg = format!(
-                                "Unused assignment or binding for local '{}'. Consider removing \
-                                 or replacing it with '_'",
-                                v_str
-                            );
-                            context.error(vec![(l.loc, msg)]);
+                            if !v_str.starts_with('_') {
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/liveness/mod.rs:297]: `v.starts_with_underscore()` would make me feel a lot less nervous here

File: language/move-lang/src/cfgir/liveness/mod.rs [hunk 2]
```diff
@@ -294,12 +294,14 @@ mod last_usage {
...
+                                let msg = format!(
+                                    "Unused assignment or binding for local '{}'. \
+                                    Consider removing or replacing it with '_'",
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/liveness/mod.rs:300]: We should update this error message to suggest prepending it with `_`

File: language/move-lang/tests/move_check/typing/bind_pop_resource.exp [hunk 3]
```diff
@@ -12,20 +12,6 @@ error:
...
- 6 │         let _r: R = R{};
-   │             ^^ Cannot ignore resource values. The value must be used
-   ·
- 6 │         let _r: R = R{};
-   │                 - The type: '0x8675309::M::R'
-   ·
- 2 │     resource struct R {}
-   │     -------- Is found to be a non-copyable type here
-   │
-
```
Comment (tnowacki) [inline language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: This should still be an error. what's going on here?

Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: Or at least there should be a new error that says "unused resource in local _r"

File: language/move-lang/src/cfgir/liveness/mod.rs (hunk 2 above)
Comment (mengxu-fb) [reply language/move-lang/src/cfgir/liveness/mod.rs:300]: Sure, I'll update it. This will lead to several updates to the `exp` files too.

File: language/move-lang/tests/move_check/typing/bind_pop_resource.exp (hunk 3 above)
Comment (mengxu-fb) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: You are right. This should still be an error.

If I just put
```
module M {
    resource struct R {}

    fun t0() {
        let _r: R = R{};
    }
}
```

It gives
```
error:

   ┌── test.move:5:24 ───
   │
 5 │         let _r: R = R{};
   │                        ^ Invalid return
   ·
 5 │         let _r: R = R{};
   │             -- The local '_r' still contains a resource value due to this assignment. The resource must be consumed before the function returns
   │

```

But if I put `let _: R = R{};` above or below `let _r`, the error is gone...

Please allow me a bit more time to check this.

Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: I think it's because I just rely on unused assignments to check that. You might need another case for:
- unused _ variables that are resourceful

Comment (mengxu-fb) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: I guess things might be a bit more complex there, even without the underscore, the `r` is not warned.
```
module M {
    resource struct R {}

    fun t0() {
        let _: R = R{};
        let r: R = R{};
    }
}
```

Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: That's very bizarre... If you need help debugging this let me know

Comment (mengxu-fb) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: My suspicion is that the pass `LocalsSafety` is shortcut by unused assignment but let me look more into this. I think I can debug this one.

Comment (mengxu-fb) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]:
```
        PassResult::Naming(nprog, errors) => {
            let (tprog, errors) = typing::translate::program(nprog, errors);
            check_errors(errors)?;
            ^^^^^^^^^^^^^^^^^^^^^^
            run(PassResult::Typing(tprog), until)
        }
        PassResult::Typing(tprog) => {
            let (hprog, errors) = hlir::translate::program(tprog);
            run(PassResult::HLIR(hprog, errors), until)
        }
        PassResult::HLIR(hprog, errors) => {
            let (cprog, errors) = cfgir::translate::program(errors, hprog);
            check_errors(errors)?;
            run(PassResult::CFGIR(cprog), until)
        }
        PassResult::CFGIR(cprog) => {
            let compiled_units = to_bytecode::translate::program(cprog)?;
            assert!(until == Pass::Compilation);
            run(PassResult::Compilation(compiled_units), Pass::Compilation)
        }
```

This `check_error()` short-circuited the execution such that the CFGIR (where `LocalSafety` lives) is not executed.

Removing it brings back the error, but I doubt that this is the right way to fix it
```
error:

   ┌── test.move:5:13 ───
   │
 5 │         let r: R = R{};
   │             ^ Unused assignment or binding for local 'r'. Consider removing it or replacing/prefixing it with '_'
   │

error:

   ┌── test.move:6:16 ───
   │
 6 │     let _: R = R{};
   │                   ^ Invalid return
   ·
 5 │         let r: R = R{};
   │             - The local 'r' still contains a resource value due to this assignment. The resource must be consumed before the function returns
   │

```

Comment (mengxu-fb) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: This might not be easy to fix, as reasoning whether a local is actually used or not needs CFGIR, but reasoning whether a `_` binds to a resource just needs `typing`.

This issue no impact on the overall correctness of the compiler. But it will leave certain errors (e.g., those checks that requires HLIR, CFGIR) undetected if there are naming or typing isssues.

Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: Oh if its just an issue with that test its fine. It is well known that you do not get unused resource warnings until after typing has completed. This is normal and expected.

But you will still need to fix that one case for resources in that one test as I mentioned above

Comment (mengxu-fb) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: I think this is also the reason why that error is no longer showing in the `bind_pop_resource.move` case (i.e., the case you mentioned above)
```
$cat bind_pop_resource.move
module M {
    resource struct R {}

    fun t0() {
        let _: R = R{};
        let _r: R = R{};
        let (_, _):(R, R) = (R{}, R{});
    }
}
```

After making `_r` as binding instead of ignore, that error will require CFGIR to check while the others did not pass the `typing` pass. Hence the original error is not printed.

I think it's better for me to put a comment in `bind_pop_resource.move` to mark that this `let _r` is also wrong, but its error won't be shown. Or better still, create another test in `locals` with only `let _r: R = R{};` in `fun t0()`.

Comment (tnowacki) [reply language/move-lang/tests/move_check/typing/bind_pop_resource.exp:28]: I would do both of those things, put a comment and make another test. Thanks!

**Comment by mengxu-fb** [comment]
Rebased past the renaming. @tnowacki would you mind having a final look on the PR?

File: language/move-lang/src/cfgir/liveness/mod.rs [hunk 4]
```diff
@@ -294,12 +294,20 @@ mod last_usage {
...
-                            let msg = format!(
-                                "Unused assignment or binding for local '{}'. Consider removing \
-                                 or replacing it with '_'",
-                                v_str
-                            );
-                            context.error(vec![(l.loc, msg)]);
+                            if !v.starts_with_underscore() {
+                                let verb = if context.is_resourceful(&v) {
+                                    // do not suggest prefixing for resource bindings
+                                    "replacing"
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/liveness/mod.rs:300]: I think this is unnecessary, replacing it with `_` if it is resourceful will result in an error, so it might be okay to just say "replacing / prefixing" always. 

I might be a bit more explicit about prefixing so maybe:

"Unused assignment or binding for local '{:v_str}'. Consider removing, replacing with '_', or prefixing with '_' e.g. '_{:v_str}'"

That might be a bit wordy, but I think it is good to squeeze in the prefix example somehow

**Comment by mengxu-fb** [comment]
Rebased to master.

**Review by tnowacki — APPROVED** [review]

**Comment by mengxu-fb** [comment]
Rebased, see if this may avoid the Circle CI glitch (2xlarge docker).
